<a href="https://colab.research.google.com/github/asadbek11211/Pandas_exercises/blob/main/_groupby().ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guruhlash. .groupby()

.groupby() metodi yordamida jadvaldagi ma'lumotlarni biror ustun qiymaltlari guruhlarga ajratib tahlil qilishimiz mumkin

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/automobile_data.csv", index_col=0)
df.head()

company   body-style  wheel-base  length engine-type  \
index                                                             
0      alfa-romero  convertible        88.6   168.8        dohc   
1      alfa-romero  convertible        88.6   168.8        dohc   
2      alfa-romero    hatchback        94.5   171.2        ohcv   
3             audi        sedan        99.8   176.6         ohc   
4             audi        sedan        99.4   176.6         ohc   

      num-of-cylinders  horsepower  average-mileage    price  
index                                                         
0                 four         111               21  13495.0  
1                 four         111               21  16500.0  
2                  six         154               19  16500.0  
3                 four         102               24  13950.0  
4                 five         115               18  17450.0

In [ ]:
df.company.values

array(['alfa-romero', 'alfa-romero', 'alfa-romero', 'audi', 'audi',
       'audi', 'audi', 'bmw', 'bmw', 'bmw', 'bmw', 'bmw', 'bmw',
       'chevrolet', 'chevrolet', 'chevrolet', 'dodge', 'dodge', 'honda',
       'honda', 'honda', 'isuzu', 'isuzu', 'isuzu', 'jaguar', 'jaguar',
       'jaguar', 'mazda', 'mazda', 'mazda', 'mazda', 'mazda',
       'mercedes-benz', 'mercedes-benz', 'mercedes-benz', 'mercedes-benz',
       'mitsubishi', 'mitsubishi', 'mitsubishi', 'mitsubishi', 'nissan',
       'nissan', 'nissan', 'nissan', 'nissan', 'porsche', 'porsche',
       'porsche', 'toyota', 'toyota', 'toyota', 'toyota', 'toyota',
       'toyota', 'toyota', 'volkswagen', 'volkswagen', 'volkswagen',
       'volkswagen', 'volvo', 'volvo'], dtype=object)

Yuqoridagi jadvalda turli avtomobil brandlari haqida ma'lumotlar jamlangan. Keling boshlanishiga har bir brand uchun ustunlarning o'rtacha qiymatini topamiz:

In [ ]:
df.groupby('company').mean()

wheel-base      length  horsepower  average-mileage  \
company                                                              
alfa-romero     90.566667  169.600000  125.333333        20.333333   
audi           101.200000  180.800000  109.250000        20.000000   
bmw            103.433333  185.033333  144.833333        19.000000   
chevrolet       92.466667  151.933333   62.666667        41.000000   
dodge           93.700000  157.300000   68.000000        31.000000   
honda           96.500000  167.200000   92.333333        26.333333   
isuzu           94.433333  160.833333   72.666667        33.333333   
jaguar         109.333333  196.966667  204.666667        14.333333   
mazda           95.900000  164.260000   75.400000        28.000000   
mercedes-benz  113.225000  197.275000  153.500000        18.000000   
mitsubishi      95.000000  164.850000   78.000000        29.500000   
nissan          95.680000  170.140000   82.800000        31.400000   
porsche         92.466667  171.166667  234.000000        17.000000   
toyota          96.957143  167.571429   75.428571        28.714286   
volkswagen      97.300000  171.700000   72.250000        31.750000   
volvo          104.300000  188.800000  114.000000        23.000000   

                      price  
company                      
alfa-romero    15498.333333  
audi           16392.500000  
bmw            27213.333333  
chevrolet       6007.000000  
dodge           6303.000000  
honda          10195.000000  
isuzu           6785.000000  
jaguar         34600.000000  
mazda           9654.800000  
mercedes-benz  35040.000000  
mitsubishi      6689.000000  
nissan          8289.000000  
porsche        35528.000000  
toyota          8216.857143  
volkswagen      8435.000000  
volvo          13177.500000

Yuqoridagi natijani tahlil qilsak, `alfa-romero` mashinalarning o'rtacha narxi (`price`) `15498$`, `honda` mashinalarining o'rtacha ot kuchi (`horsepower`) `92.3` ga teng ekan.

Bitta ustun uchun qiymatlarni olish ham mumkin:

In [ ]:
df.groupby('company').price.min()

company
alfa-romero      13495.0
audi             13950.0
bmw              16430.0
chevrolet         5151.0
dodge             6229.0
honda             7295.0
isuzu             6785.0
jaguar           32250.0
mazda             5195.0
mercedes-benz    25552.0
mitsubishi        5389.0
nissan            6649.0
porsche          34028.0
toyota            5348.0
volkswagen        7775.0
volvo            12940.0
Name: price, dtype: float64

In [ ]:
df.groupby(['company','num-of-cylinders']).mean()

wheel-base      length  horsepower  \
company       num-of-cylinders                                       
alfa-romero   four               88.600000  168.800000  111.000000   
              six                94.500000  171.200000  154.000000   
audi          five              101.666667  182.200000  111.666667   
              four               99.800000  176.600000  102.000000   
bmw           four              101.200000  176.800000  101.000000   
              six               104.550000  189.150000  166.750000   
chevrolet     four               94.500000  157.350000   70.000000   
              three              88.400000  141.100000   48.000000   
dodge         four               93.700000  157.300000   68.000000   
honda         four               96.500000  167.200000   92.333333   
isuzu         four               94.433333  160.833333   72.666667   
jaguar        six               113.000000  199.600000  176.000000   
              twelve            102.000000  191.700000  262.000000   
mazda         four               96.050000  163.075000   69.000000   
              two                95.300000  169.000000  101.000000   
mercedes-benz eight             116.450000  203.650000  184.000000   
              five              110.000000  190.900000  123.000000   
mitsubishi    four               95.000000  164.850000   78.000000   
nissan        four               94.500000  166.525000   65.500000   
              six               100.400000  184.600000  152.000000   
porsche       eight              98.400000  175.700000  288.000000   
              six                89.500000  168.900000  207.000000   
toyota        four               95.700000  164.200000   62.000000   
              six               104.500000  187.800000  156.000000   
volkswagen    four               97.300000  171.700000   72.250000   
volvo         four              104.300000  188.800000  114.000000   

                                average-mileage         price  
company       num-of-cylinders                                 
alfa-romero   four                    21.000000  14997.500000  
              six                     19.000000  16500.000000  
audi          five                    18.666667  17206.666667  
              four                    24.000000  13950.000000  
bmw           four                    23.000000  16677.500000  
              six                     17.000000  32481.250000  
chevrolet     four                    38.000000   6435.000000  
              three                   47.000000   5151.000000  
dodge         four                    31.000000   6303.000000  
honda         four                    26.333333  10195.000000  
isuzu         four                    33.333333   6785.000000  
jaguar        six                     15.000000  33900.000000  
              twelve                  13.000000  36000.000000  
mazda         four                    30.750000   9107.250000  
              two                     17.000000  11845.000000  
mercedes-benz eight                   14.000000  43180.000000  
              five                    22.000000  26900.000000  
mitsubishi    four                    29.500000   6689.000000  
nissan        four                    34.500000   6986.500000  
              six                     19.000000  13499.000000  
porsche       eight                   17.000000           NaN  
              six                     17.000000  35528.000000  
toyota        four                    30.333333   6961.333333  
              six                     19.000000  15750.000000  
volkswagen    four                    31.750000   8435.000000  
volvo         four                    23.000000  13177.500000

.groupby() yordamida nafaqat o'rta balki boshqa qiymatlarni ham hisblash mumkin

In [ ]:
df.groupby('company').max()

body-style  wheel-base  length engine-type num-of-cylinders  \
company                                                                     
alfa-romero    hatchback        94.5   171.2        ohcv              six   
audi               wagon       105.8   192.7         ohc             four   
bmw                sedan       110.0   197.0         ohc              six   
chevrolet          sedan        94.5   158.8         ohc            three   
dodge          hatchback        93.7   157.3         ohc             four   
honda              wagon        96.5   175.4         ohc             four   
isuzu              sedan        94.5   170.7         ohc             four   
jaguar             sedan       113.0   199.6        ohcv           twelve   
mazda              sedan       104.9   175.0       rotor              two   
mercedes-benz      wagon       120.9   208.1        ohcv             five   
mitsubishi         sedan        96.3   172.4         ohc             four   
nissan             wagon       100.4   184.6        ohcv              six   
porsche        hatchback        98.4   175.7        ohcf              six   
toyota             wagon       104.5   187.8         ohc              six   
volkswagen         sedan        97.3   171.7         ohc             four   
volvo              wagon       104.3   188.8         ohc             four   

               horsepower  average-mileage    price  
company                                              
alfa-romero           154               21  16500.0  
audi                  115               24  18920.0  
bmw                   182               23  41315.0  
chevrolet              70               47   6575.0  
dodge                  68               31   6377.0  
honda                 101               30  12945.0  
isuzu                  78               38   6785.0  
jaguar                262               15  36000.0  
mazda                 101               31  18344.0  
mercedes-benz         184               22  45400.0  
mitsubishi             88               37   8189.0  
nissan                152               45  13499.0  
porsche               288               17  37028.0  
toyota                156               35  15750.0  
volkswagen            100               37   9995.0  
volvo                 114               23  13415.0

.agg() motodi yordamida birdaniga bir nechta statik ma'lumotlarni olishimiz ham mumkin:

In [ ]:
df.groupby('company').agg([len, max, min, lambda x: x*2]).head()

body-style                          \
                   len        max          min   
company                                          
alfa-romero          3  hatchback  convertible   
audi                 4      wagon        sedan   
bmw                  6      sedan        sedan   
chevrolet            3      sedan    hatchback   
dodge                2  hatchback    hatchback   

                                                               engine-type  \
                                                    <lambda_0>         len   
company                                                                      
alfa-romero  [convertibleconvertible, convertibleconvertibl...           3   
audi          [sedansedan, sedansedan, sedansedan, wagonwagon]           4   
bmw          [sedansedan, sedansedan, sedansedan, sedanseda...           6   
chevrolet    [hatchbackhatchback, hatchbackhatchback, sedan...           3   
dodge                 [hatchbackhatchback, hatchbackhatchback]           2   

                                                                           \
              max   min                                        <lambda_0>   
company                                                                     
alfa-romero  ohcv  dohc                    [dohcdohc, dohcdohc, ohcvohcv]   
audi          ohc   ohc                  [ohcohc, ohcohc, ohcohc, ohcohc]   
bmw           ohc   ohc  [ohcohc, ohcohc, ohcohc, ohcohc, ohcohc, ohcohc]   
chevrolet     ohc     l                              [ll, ohcohc, ohcohc]   
dodge         ohc   ohc                                  [ohcohc, ohcohc]   

            num-of-cylinders               \
                         len    max   min   
company                                     
alfa-romero                3    six  four   
audi                       4   four  five   
bmw                        6    six  four   
chevrolet                  3  three  four   
dodge                      2   four  four   

                                                                
                                                    <lambda_0>  
company                                                         
alfa-romero                       [fourfour, fourfour, sixsix]  
audi                  [fourfour, fivefive, fivefive, fivefive]  
bmw          [fourfour, fourfour, sixsix, sixsix, sixsix, s...  
chevrolet                     [threethree, fourfour, fourfour]  
dodge                                     [fourfour, fourfour]

In [ ]:
df.groupby('company').agg([np.mean, max, len]).head()

wheel-base                 length             horsepower       \
                   mean    max len        mean    max len        mean  max   
company                                                                      
alfa-romero   90.566667   94.5   3  169.600000  171.2   3  125.333333  154   
audi         101.200000  105.8   4  180.800000  192.7   4  109.250000  115   
bmw          103.433333  110.0   6  185.033333  197.0   6  144.833333  182   
chevrolet     92.466667   94.5   3  151.933333  158.8   3   62.666667   70   
dodge         93.700000   93.7   2  157.300000  157.3   2   68.000000   68   

                average-mileage                 price               
            len            mean max len          mean      max len  
company                                                             
alfa-romero   3       20.333333  21   3  15498.333333  16500.0   3  
audi          4       20.000000  24   4  16392.500000  18920.0   4  
bmw           6       19.000000  23   6  27213.333333  41315.0   6  
chevrolet     3       41.000000  47   3   6007.000000   6575.0   3  
dodge         2       31.000000  31   2   6303.000000   6377.0   2